// investing-economic-calendar/dist/index.js
"use strict";
// import { extractEconomicEvents } from './crawler';
// import { CalendarType } from './models/CalendarType';
// import { Country } from './models/Country';
// import { Importance } from './models/Importance';
// import { Language } from './models/Language';
// import { TimeZone } from './models/TimeZone';
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchEconomicEvents = exports.TimeZone = exports.Language = exports.Importance = exports.Country = exports.CalendarType = void 0;
// (async () => {
// 	const events = await extractEconomicEvents({
// 		importance: [Importance.HIGH],
// 		calType:    CalendarType.DAILY,
// 		lang:       Language.FRENCH,
// 		timeZone:   TimeZone.EASTERN_TIME_US_CANADA,
// 		countries:  [
// 			Country.AUSTRALIA, Country.CANADA, Country.EURO_ZONE, Country.FRANCE,
// 			Country.GERMANY, Country.JAPAN, Country.NEW_ZEALAND, Country.SWITZERLAND,
// 			Country.UNITED_KINGDOM, Country.UNITED_STATES],
// 	});
// 	console.log(events);
// })();
var CalendarType_1 = require("./models/CalendarType");
Object.defineProperty(exports, "CalendarType", { enumerable: true, get: function () { return CalendarType_1.CalendarType; } });
var Country_1 = require("./models/Country");
Object.defineProperty(exports, "Country", { enumerable: true, get: function () { return Country_1.Country; } });
var Importance_1 = require("./models/Importance");
Object.defineProperty(exports, "Importance", { enumerable: true, get: function () { return Importance_1.Importance; } });
var Language_1 = require("./models/Language");
Object.defineProperty(exports, "Language", { enumerable: true, get: function () { return Language_1.Language; } });
var TimeZone_1 = require("./models/TimeZone");
Object.defineProperty(exports, "TimeZone", { enumerable: true, get: function () { return TimeZone_1.TimeZone; } });
var crawler_1 = require("./crawler");
Object.defineProperty(exports, "fetchEconomicEvents", { enumerable: true, get: function () { return crawler_1.fetchEconomicEvents; } });
